syntax = "proto3";

option csharp_namespace = "FunnelCloud.Agent.Grpc";

package funnelcloud;

// Task execution service for FunnelCloud agents
// All calls require mTLS with CA-pinned certificates
service TaskService {
    // Execute a task on this agent
    rpc Execute (TaskRequest) returns (TaskResult);
    
    // Execute a task with streaming output
    rpc ExecuteStreaming (TaskRequest) returns (stream TaskOutput);
    
    // Get the status of a running task
    rpc GetStatus (TaskStatusRequest) returns (TaskStatusResponse);
    
    // Cancel a running task
    rpc Cancel (CancelRequest) returns (CancelResponse);
    
    // Ping to verify connection
    rpc Ping (PingRequest) returns (PingResponse);
}

// Task execution request
message TaskRequest {
    string task_id = 1;
    TaskType type = 2;
    string command = 3;
    int32 timeout_seconds = 4;
    bool require_elevation = 5;
    string working_directory = 6;
    map<string, string> environment = 7;
}

enum TaskType {
    TASK_TYPE_UNSPECIFIED = 0;
    SHELL = 1;           // Native shell (cmd/bash)
    POWERSHELL = 2;      // PowerShell Core
    READ_FILE = 3;       // Read file contents
    WRITE_FILE = 4;      // Write file contents
    LIST_DIRECTORY = 5;  // List directory contents
    DOTNET_CODE = 6;     // Execute .NET code snippet
}

// Task execution result
message TaskResult {
    bool success = 1;
    string stdout = 2;
    string stderr = 3;
    int32 exit_code = 4;
    ErrorCode error_code = 5;
    int64 duration_ms = 6;
    string task_id = 7;
}

enum ErrorCode {
    ERROR_CODE_UNSPECIFIED = 0;
    ERROR_NONE = 1;
    ERROR_TIMEOUT = 2;
    ERROR_ELEVATION_REQUIRED = 3;
    ERROR_NOT_FOUND = 4;
    ERROR_PERMISSION_DENIED = 5;
    ERROR_INTERNAL = 6;
    ERROR_CANCELLED = 7;
}

// Streaming output during task execution
message TaskOutput {
    string task_id = 1;
    OutputType type = 2;
    string content = 3;
    int64 timestamp_ms = 4;
}

enum OutputType {
    OUTPUT_TYPE_UNSPECIFIED = 0;
    STDOUT = 1;
    STDERR = 2;
    STATUS = 3;
}

// Task status request
message TaskStatusRequest {
    string task_id = 1;
}

message TaskStatusResponse {
    string task_id = 1;
    TaskState state = 2;
    int32 progress_percent = 3;
    string status_message = 4;
}

enum TaskState {
    TASK_STATE_UNSPECIFIED = 0;
    TASK_PENDING = 1;
    TASK_RUNNING = 2;
    TASK_COMPLETED = 3;
    TASK_FAILED = 4;
    TASK_CANCELLED = 5;
}

// Cancel request
message CancelRequest {
    string task_id = 1;
    bool force = 2;  // Force kill if graceful cancel fails
}

message CancelResponse {
    bool cancelled = 1;
    string message = 2;
}

// Ping for connection verification
message PingRequest {
    int64 timestamp_ms = 1;
}

message PingResponse {
    int64 request_timestamp_ms = 1;
    int64 response_timestamp_ms = 2;
    string agent_id = 3;
    string agent_version = 4;
}
